cmd/lotus-shed/fip-0036.go

```diff
@@ -0,0 +1,565 @@
+		if err != nil {
+			return xerrors.Errorf("failed to get voters: ", err)
+		}
+
+		type minerBriefInfo struct {
+			rawBytePower abi.StoragePower
+			dealPower    abi.StoragePower
+			balance      abi.TokenAmount
+		}
+
+		// power actor
+		pa, err := st.GetActor(power.Address)
+		if err != nil {
+			return xerrors.Errorf("failed to get power actor: \n", err)
+		}
+
+		powerState, err := power.Load(store, pa)
+		if err != nil {
+			return xerrors.Errorf("failed to get power state: \n", err)
+		}
+
+		//market actor
+		ma, err := st.GetActor(market.Address)
+		if err != nil {
+			return xerrors.Errorf("fail to get market actor: ", err)
+		}
+
+		marketState, err := market.Load(store, ma)
+		if err != nil {
+			return xerrors.Errorf("fail to load market state: ", err)
+		}
+
+		//init actor
+		ia, err := st.GetActor(_init.Address)
+		if err != nil {
+			return xerrors.Errorf("fail to get init actor: ", err)
+		}
+
+		initState, err := _init.Load(store, ia)
+		if err != nil {
+			return xerrors.Errorf("fail to load init state: ", err)
+		}
+
+		initMap, err := initState.AddressMap()
+		if err != nil {
+			return xerrors.Errorf("fail to load init map: ", err)
+		}
+
+		lookupId := func(addr address.Address) address.Address {
+			if addr.Protocol() == address.ID {
+				return addr
+			}
+
+			var actorID cbg.CborInt
+			if found, err := initMap.Get(abi.AddrKey(addr), &actorID); err != nil {
+				panic(err)
+			} else if found {
+				// Reconstruct address from the ActorID.
+				idAddr, err := address.NewIDAddress(uint64(actorID))
+				if err != nil {
+					panic(err)
+				}
+				return idAddr
+			}
+
+			panic("didn't find addr")
+		}
+
+		// we need to build several pieces of information, as we traverse the state tree:
+		// a map of accounts to every msig that they are a signer of
+		accountsToMultisigs := make(map[address.Address][]address.Address)
+		// a map of multisigs to some info about them for quick lookup
+		msigActorsInfo := make(map[address.Address]msigBriefInfo)
+		// a map of accounts to every miner that they are an owner of
+		ownerMap := make(map[address.Address][]address.Address)
+		// a map of accounts to every miner that they are a worker of
+		workerMap := make(map[address.Address][]address.Address)
+		// a map of miners to some info aboout them for quick lookup
+		minerActorsInfo := make(map[address.Address]minerBriefInfo)
+		// a map of client addresses to deal data stored in proposals
+		clientToDealStorage := make(map[address.Address]abi.StoragePower)
+
+		fmt.Println("iterating over all actors")
+		count := 0
+		err = st.ForEach(func(addr address.Address, act *types.Actor) error {
+			if count%200000 == 0 {
+				fmt.Println("processed ", count, " actors building maps")
+			}
+			count++
+			if builtin.IsMultisigActor(act.Code) {
+				ms, err := multisig.Load(store, act)
+				if err != nil {
+					return fmt.Errorf("load msig failed %v", err)
+
+				}
+
+				// TODO: Confirm that these are always ID addresses
```

Comment:
Confirmed.

Reply:
Oh yeah, i resolved them anyway further down...thank you!